File: windows/Transcribe.Core/Transcriber.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace Transcribe.Core;

/// <summary>
/// Whisper HTTP client + the silence-hallucination filter + the best-effort per-track loop.
/// Direct port of meetrec.py <c>transcribe_track_whisper</c> and Swift
/// <c>transcribeChunk</c> / <c>transcribeTrack</c>. The multipart contract is hand-built to
/// match the macOS app byte-for-byte: <c>model=whisper-1</c>, <c>response_format=verbose_json</c>,
/// the literal <c>timestamp_granularities[]</c> field, <c>language</c> only when set, and the
/// file part with an <c>audio/mp4</c> Content-Type.
/// </summary>
public sealed class Transcriber
{
    private readonly HttpClient _http;

    /// <summary>verbose_json file part content type, matching the Swift app's "audio/mp4".</summary>
    public const string FileContentType = "audio/mp4";

    /// <summary>The only OpenAI model that returns per-segment timestamps.</summary>
    public const string TimestampModel = "whisper-1";

    /// <summary>Chunks smaller than this are treated as empty and skipped (Swift: 256 bytes).</summary>
    public const int MinChunkBytes = 256;

    public Transcriber(HttpClient http)
    {
        _http = http;
    }

    /// <summary>
    /// True if this segment is a whisper silence-hallucination and must be dropped:
    /// <c>no_speech_prob &gt; 0.6 &amp;&amp; avg_logprob &lt; -1.0</c>. Pure; unit-tested.
    /// Absent probabilities default to 0.0 (so the filter never fires on them), matching the
    /// Python <c>getattr(..., 0.0) or 0.0</c> and the Swift <c>?? 0.0</c>.
    /// </summary>
    public static bool IsSilenceHallucination(WhisperSegment s)
    {
        var nsp = s.NoSpeechProb ?? 0.0;
        var alp = s.AvgLogprob ?? 0.0;
        return nsp > 0.6 && alp < -1.0;
    }

    /// <summary>
    /// Build the multipart form for one chunk. Exposed (internal) so a test can assert the exact
    /// field set / ordering / content-types without a live HTTP call.
    /// </summary>
    internal static MultipartFormDataContent BuildForm(byte[] fileBytes, string fileName, string? language)
    {
        // Fixed boundary keeps the request deterministic and testable; the proxy doesn't care
        // about the boundary value, only that it's consistent with the header.
        var boundary = "Boundary-" + Guid.NewGuid().ToString("N");
        var form = new MultipartFormDataContent(boundary);

        form.Add(StringField("whisper-1"), "model");
        form.Add(StringField("verbose_json"), "response_format");
        form.Add(StringField("segment"), "timestamp_granularities[]"); // literal brackets, one field
        if (!string.IsNullOrEmpty(language))
        {
            form.Add(StringField(language), "language");
        }

        var fileContent = new ByteArrayContent(fileBytes);
        fileContent.Headers.ContentType = new MediaTypeHeaderValue(FileContentType);
        form.Add(fileContent, "file", fileName);

        return form;

        static StringContent StringField(string value)
        {
            // No charset on simple fields, matching the hand-built macOS multipart parts.
            var c = new StringContent(value);
            c.Headers.ContentType = null;
            return c;
        }
    }

    /// <summary>
    /// One chunk → whisper-1 verbose_json. Throws on non-2xx (the caller continues to the next
    /// chunk, best-effort). Port of Swift <c>transcribeChunk</c>.
    /// </summary>
    public async Task<VerboseTranscription> TranscribeChunkAsync(
        OpenAIConfig config, byte[] fileBytes, string fileName, string? language,
        CancellationToken ct = default)
    {
        using var req = new HttpRequestMessage(HttpMethod.Post,
            CombineUri(config.BaseUrl, "audio/transcriptions"));
        req.Headers.TryAddWithoutValidation(config.AuthName, config.AuthValue);
        req.Content = BuildForm(fileBytes, fileName, language);

        using var resp = await _http.SendAsync(req, ct).ConfigureAwait(false);
        var body = await resp.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
        if (!resp.IsSuccessStatusCode)
        {
            var snippet = body.Length > 400 ? body[..400] : body;
            throw new HttpRequestException($"HTTP {(int)resp.StatusCode}: {snippet}");
        }

        var parsed = JsonSerializer.Deserialize<VerboseTranscription>(body)
                     ?? new VerboseTranscription();
        return parsed;
    }

    /// <summary>
    /// Result of transcribing one track: its kept segments plus the next global idx so the
    /// caller keeps a single input-order counter (mic first, then system) for the merge tiebreak.
    /// </summary>
    public readonly record struct TrackResult(IReadOnlyList<Segment> Segments, int NextIdx);

    /// <summary>
    /// Port of meetrec.py <c>transcribe_track_whisper</c> / Swift <c>transcribeTrack</c>.
    /// Per-chunk failures CONTINUE (best-effort partial transcript), logged via
    /// <paramref name="log"/>. Drops empty text and silence-hallucinations, applies the chunk
    /// offset + baseOffset to globalize timestamps, clamps end ≥ start, and assigns a global,
    /// monotonically increasing idx.
    /// </summary>
    public async Task<TrackResult> TranscribeTrackAsync(
        OpenAIConfig config,
        IReadOnlyList<AudioChunk> chunks,
        double baseOffset,
        string label,
        Source source,
        string? language,
        int startIdx,
        Action<string>? log = null,
        CancellationToken ct = default)
    {
        var segs = new List<Segment>();
        var idx = startIdx;

        for (var i = 0; i < chunks.Count; i++)
        {
            var chunk = chunks[i];
            long size;
            try { size = new FileInfo(chunk.Path).Length; }
            catch { size = 0; }
            if (size < MinChunkBytes)
            {
                log?.Invoke($"  skip {Path.GetFileName(chunk.Path)} (empty)");
                continue;
            }

            log?.Invoke($"Transcribing {SourceName(source)} chunk {i + 1}/{chunks.Count}…");

            VerboseTranscription resp;
            try
            {
                var bytes = await File.ReadAllBytesAsync(chunk.Path, ct).ConfigureAwait(false);
                resp = await TranscribeChunkAsync(config, bytes, Path.GetFileName(chunk.Path),
                    language, ct).ConfigureAwait(false);
            }
            catch (Exception e)
            {
                log?.Invoke($"  {SourceName(source)} chunk {i + 1} failed: {e.Message}");
                continue;
            }

            var off = baseOffset + chunk.Offset;
            foreach (var s in resp.Segments ?? new List<WhisperSegment>())
            {
                var txt = (s.Text ?? "").Trim();
                if (txt.Length == 0) continue;
                if (IsSilenceHallucination(s)) continue;
                var st = s.Start + off;
                var en = Math.Max(s.End + off, st);
                segs.Add(new Segment(st, en, label, txt, source, idx));
                idx++;
            }
        }

        return new TrackResult(segs, idx);
    }

    private static string SourceName(Source s) => s == Source.Mic ? "mic" : "system";

    /// <summary>
    /// Join base + relative, preserving the base path (e.g. ".../v1/audio/transcriptions").
    /// new Uri(base, "audio/...") would drop "/v1"; appending segments avoids that.
    /// </summary>
    private static Uri CombineUri(Uri baseUrl, string relative)
    {
        var b = baseUrl.AbsoluteUri.TrimEnd('/');
        return new Uri($"{b}/{relative}");
    }
}

File: TranscribeBar/Sources/TranscribeCore/Transcriber.swift
import Foundation

struct RuntimeError: Error, CustomStringConvertible {
    let message: String
    init(_ m: String) { message = m }
    var description: String { message }
}

extension Data {
    mutating func appendString(_ s: String) { append(s.data(using: .utf8)!) }
}

// MARK: - verbose_json schema (whisper-1). Optionals guarded per the Python contract.

public struct WhisperSegment: Decodable {
    public let start: Double
    public let end: Double
    public let text: String?
    public let avgLogprob: Double?
    public let noSpeechProb: Double?
    enum CodingKeys: String, CodingKey {
        case start, end, text
        case avgLogprob = "avg_logprob"
        case noSpeechProb = "no_speech_prob"
    }
}

public struct VerboseTranscription: Decodable {
    public let language: String?
    public let duration: Double?
    public let text: String?
    public let segments: [WhisperSegment]?
}

// MARK: - Auth (proxy base-URL + bearer, or direct key). Single provider seam.

public struct OpenAIConfig: Sendable {
    public var baseURL: URL                 // e.g. https://api.openai.com/v1  or  the proxy
    public var authName: String
    public var authValue: String
    public init(baseURL: URL, authName: String = "Authorization", authValue: String) {
        self.baseURL = baseURL
        self.authName = authName
        self.authValue = authValue
    }
    /// Dev/headless config from env: OPENAI_BASE_URL (default api.openai.com/v1) + OPENAI_API_KEY.
    public static func fromEnv() -> OpenAIConfig {
        let env = ProcessInfo.processInfo.environment
        let base = env["OPENAI_BASE_URL"] ?? "https://api.openai.com/v1"
        let key = env["OPENAI_API_KEY"] ?? ""
        return OpenAIConfig(baseURL: URL(string: base)!, authValue: "Bearer \(key)")
    }
    public var hasKey: Bool { authValue != "Bearer " && !authValue.isEmpty }
}

/// One chunk → whisper-1 verbose_json. Multipart hand-built (no SDK); note the literal
/// `timestamp_granularities[]` field name and the per-file Content-Type.
public func transcribeChunk(fileURL: URL, language: String?, config: OpenAIConfig) async throws -> VerboseTranscription {
    let boundary = "Boundary-\(UUID().uuidString)"
    var req = URLRequest(url: config.baseURL.appendingPathComponent("audio/transcriptions"))
    req.httpMethod = "POST"
    req.setValue(config.authValue, forHTTPHeaderField: config.authName)
    req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

    var body = Data()
    func field(_ name: String, _ value: String) {
        body.appendString("--\(boundary)\r\n")
        body.appendString("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n")
        body.appendString("\(value)\r\n")
    }
    field("model", "whisper-1")
    field("response_format", "verbose_json")
    field("timestamp_granularities[]", "segment")   // literal brackets, one part
    if let language, !language.isEmpty { field("language", language) }

    let fileData = try Data(contentsOf: fileURL)
    body.appendString("--\(boundary)\r\n")
    body.appendString("Content-Disposition: form-data; name=\"file\"; filename=\"\(fileURL.lastPathComponent)\"\r\n")
    body.appendString("Content-Type: audio/mp4\r\n\r\n")
    body.append(fileData)
    body.appendString("\r\n")
    body.appendString("--\(boundary)--\r\n")
    req.httpBody = body

    let (data, resp) = try await URLSession.shared.data(for: req)
    guard let http = resp as? HTTPURLResponse else { throw RuntimeError("no HTTP response") }
    guard (200..<300).contains(http.statusCode) else {
        let msg = String(data: data, encoding: .utf8) ?? "<binary>"
        throw RuntimeError("HTTP \(http.statusCode): \(msg.prefix(400))")
    }
    return try JSONDecoder().decode(VerboseTranscription.self, from: data)
}

// MARK: - Summary (chat.completions)

private struct ChatResponse: Decodable {
    struct Choice: Decodable { struct Msg: Decodable { let content: String? }; let message: Msg }
    let choices: [Choice]
}

/// Port of meetrec.py `summarize`. Best-effort; returns nil on any failure.
public func summarize(_ transcript: String, model: String = "gpt-4o", config: OpenAIConfig) async -> String? {
    let prompt = "Summarize this meeting transcript. Return: a 3-5 sentence overview, then a "
        + "bulleted list of decisions, then a bulleted list of action items with an owner if "
        + "one is mentioned.\n\n" + transcript
    var req = URLRequest(url: config.baseURL.appendingPathComponent("chat/completions"))
    req.httpMethod = "POST"
    req.setValue(config.authValue, forHTTPHeaderField: config.authName)
    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
    let body: [String: Any] = ["model": model,
                               "messages": [["role": "user", "content": prompt]]]
    req.httpBody = try? JSONSerialization.data(withJSONObject: body)
    guard let (data, resp) = try? await URLSession.shared.data(for: req),
          let http = resp as? HTTPURLResponse, (200..<300).contains(http.statusCode),
          let decoded = try? JSONDecoder().decode(ChatResponse.self, from: data),
          let text = decoded.choices.first?.message.content?.trimmingCharacters(in: .whitespacesAndNewlines),
          !text.isEmpty
    else { return nil }
    return text
}

/// Port of meetrec.py `transcribe_track_whisper`. Per-chunk failures CONTINUE (best-effort,
/// partial transcript). Returns segments + the next idx so the caller keeps a global
/// input-order counter (mic track first, then system) for the merge tiebreak.
public func transcribeTrack(chunks: [(url: URL, offset: Double)],
                            baseOffset: Double,
                            label: String,
                            source: Source,
                            language: String?,
                            config: OpenAIConfig,
                            startIdx: Int) async -> (segments: [Segment], nextIdx: Int) {
    var segs: [Segment] = []
    var idx = startIdx
    for (i, chunk) in chunks.enumerated() {
        let attrs = try? FileManager.default.attributesOfItem(atPath: chunk.url.path)
        let size = (attrs?[.size] as? Int) ?? 0
        if size < 256 {
            FileHandle.standardError.write("  skip \(chunk.url.lastPathComponent) (empty)\n".data(using: .utf8)!)
            continue
        }
        FileHandle.standardError.write("Transcribing \(source) chunk \(i + 1)/\(chunks.count)…\n".data(using: .utf8)!)
        let resp: VerboseTranscription
        do {
            resp = try await transcribeChunk(fileURL: chunk.url, language: language, config: config)
        } catch {
            FileHandle.standardError.write("  \(source) chunk \(i + 1) failed: \(error)\n".data(using: .utf8)!)
            continue
        }
        let off = baseOffset + chunk.offset
        for s in resp.segments ?? [] {
            let txt = (s.text ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            if txt.isEmpty { continue }
            let nsp = s.noSpeechProb ?? 0.0
            let alp = s.avgLogprob ?? 0.0
            if nsp > 0.6 && alp < -1.0 { continue }   // drop silence-hallucinations
            let st = s.start + off
            let en = max(s.end + off, st)
            segs.append(Segment(start: st, end: en, label: label, text: txt, source: source, idx: idx))
            idx += 1
        }
    }
    return (segs, idx)
}
